File: example/helloworld/go/helloworld.mrpc.go
package main

import (
	"encoding/json"
	"mrpc"
)

var Hello_method_names = []string{
	"/helloworld.Greeter/SayHello",
}

type SayHelloRequest struct {
	Name string `json:"name"`
}

func (r *SayHelloRequest) ToString() (string, error) {
	data, err := json.Marshal(r)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

type SayHelloResponse struct {
	Message string `json:"message"`
}

func (r *SayHelloResponse) FromString(data string) error {
	return json.Unmarshal([]byte(data), r)
}

type HelloClient struct {
	client *mrpc.Client
}

func NewHelloClient(s string) *HelloClient {
	return &HelloClient{
		client: mrpc.NewClient(s),
	}
}

func (h *HelloClient) SayHello(request *SayHelloRequest) (string, error) {
	response := &SayHelloResponse{}
	err := h.client.Send(Hello_method_names[0], request, response)
	return response.Message, err
}

func (h *HelloClient) AsyncSayHello(request *SayHelloRequest) (string, error) {
	return h.client.AsyncSend(Hello_method_names[0], request)
}

func (h *HelloClient) CallbackSayHello(request *SayHelloRequest, callback func(string, error)) {
	response := &SayHelloResponse{}
	h.client.CallbackSend(Hello_method_names[0], request, response, func(err error) {
		callback(response.Message, err)
	})
}

func (h *HelloClient) Receive(key string) (string, error) {
	response := &SayHelloResponse{}
	err := h.client.Receive(key, response)
	return response.Message, err
}

func (h *HelloClient) Close() {
	h.client.Close()
}

File: example/helloworld/python/helloworld_mrpc/helloworld_mrpc.py
import mrpc
import json
from typing import Callable, Optional

Callback = Callable[[str, Exception | None], None]


Greeter_METHOD_NAMES = [
    "/helloworld.Greeter/SayHello",
]


class SayHelloRequest(mrpc.ParseToJson):
    def __init__(self, name: Optional[str] = None):
        self.name = name

    def toString(self) -> str:
        return json.dumps({"name": self.name})
    
    def fromString(self, data: str):
        obj = json.loads(data)
        self.name = obj.get("name", "")


class SayHelloResponse(mrpc.ParseFromJson):
    def __init__(self):
        self.message = ""

    def toString(self) -> str:
        return json.dumps({"message": self.message})

    def fromString(self, data: str):
        obj = json.loads(data)
        self.message = obj.get("message", "")



class GreeterClient(mrpc.Client):
    def __init__(self, server_address: str):
        super().__init__(server_address)

    def SayHello(self, request: SayHelloRequest) -> tuple[str, Exception | None]:
        response = SayHelloResponse()
        err = super().Send(Greeter_METHOD_NAMES[0], request, response)
        return response.message, err

    def AsyncSayHello(self, request: SayHelloRequest) -> tuple[str, Exception | None]:
        return super().AsyncSend(Greeter_METHOD_NAMES[0], request)

    def CallbackSayHello(self, request: SayHelloRequest, callback: Callback):
        response = SayHelloResponse()
        super().CallbackSend(
            Greeter_METHOD_NAMES[0],
            request,
            response,
            lambda err: callback(response.message, err),
        )

    def ReceiveSayHello(self, key: str) -> tuple[str, Exception | None]:
        response = SayHelloResponse()
        err = super().Receive(key, response)
        return response.message, err


class GreeterService(mrpc.MrpcService):
    def __init__(self):
        super().__init__("helloworld.Greeter")
        self.AddHandler(
            Greeter_METHOD_NAMES[0], SayHelloRequest, SayHelloResponse, 
            lambda request, response: self.SayHello(request, response)
        )

    def SayHello(self, request: 'SayHelloRequest', response: 'SayHelloResponse') -> mrpc.MrpcError | None:
        pass


class GreeterServer(mrpc.Server):
    def __init__(self, server_address: str):
        super().__init__(server_address)